### PR TITLE
implementation of a finite MPS class

### PR DESCRIPTION
**Pull request by mganahl**

### CODE REVIEW

File: mps/mps.py [hunk 1]
```diff
@@ -0,0 +1,574 @@
+# Copyright 2019 The TensorNetwork Authors
```
Comment (Thenerdstation) [inline mps/mps.py]: Put all of this under `tensornetwork/`

File: tensornetwork/__init__.py [hunk 2]
```diff
@@ -6,6 +6,7 @@
 from tensornetwork.visualization.graphviz import to_graphviz
 from tensornetwork import contractors
 from tensornetwork import config
+from mps import mps
```
Comment (Thenerdstation) [inline tensornetwork/__init__.py]: Move the code to `tensornetwork/` and change this to `from tensornetwork.mps import mps`

File: mps/mps_test.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+import tensorflow as tf
+
+from jax.config import config
+
+config.update("jax_enable_x64", True)
+tf.compat.v1.enable_v2_behavior()
+
+
+def get_random_np(shape, dtype):
```
Comment (MichaelMarien) [inline mps/mps_test.py]: For all the random operations, I suggest to add a fixed seed to ensure reproducibility of tests

File: mps/mps.py [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+        if site - 1 in sites:
+          right_env.reorder_edges([edges[site], conj_edges[site]])
+          right_envs[site - 1] = right_env.tensor
+
+    return right_envs
+
+  def __len__(self):
+    return len(self.nodes)
+
+  def transfer_operator(self, site: int, direction: Union[Text, int],
```
Comment (MichaelMarien) [inline mps/mps.py]: just thinking ahead, for infinite MPS the concept of transfer matrix makes sense as an object (although for efficiency, one might not want to construct it). For the sake of. a common API ,perhaps rename this to `apply_transfer_matrix`?

File: tensornetwork/__init__.py (hunk 2 above)
Comment (mganahl) [reply tensornetwork/__init__.py]: This seems to create a cyclic import (linter is complaining about this)

File: mps/mps_test.py (hunk 3 above)
Comment (Thenerdstation) [reply mps/mps_test.py]: We've been burned by this before.

File: tensornetwork/matrixproductstates/mps.py [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+import numpy as np
+from typing import Any, List, Optional, Text, Type, Union, Dict
+Tensor = Any
+
+
+class FiniteMPS(TensorNetwork):
+  """
+  An MPS class for finite systems.
+  `FiniteMPS` keeps track of the nodes of the network by storing them in a list
+  `FiniteMPS.nodes`. Any external changes to this list will potentially corrupt
```
Comment (Thenerdstation) [inline tensornetwork/matrixproductstates/mps.py]: You can make that a property to avoid any modifications. Sure, someone could always modify `FiniteMPS._nodes` too, but then that is their own fault. :)

File: tensornetwork/matrixproductstates/mps.py [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+    ]  #redundant?!
+    for site in range(len(self.nodes) - 1):
+      self.connect(self.nodes[site][2], self.nodes[site + 1][0])
+    self.center_position = center_position
+
+  def save(self, path: str):
+    raise NotImplementedError()
+
+  @property
+  def D(self) -> List:
```
Comment (Thenerdstation) [inline tensornetwork/matrixproductstates/mps.py]: Give this a full name

File: tensornetwork/matrixproductstates/mps.py [hunk 7]
```diff
@@ -0,0 +1,564 @@
...
+    """
+    Return a list of bond dimensions of FiniteMPS
+    """
+    return [self.nodes[0].shape[0]] + [node.shape[2] for node in self.nodes]
+
+  @property
+  def d(self) -> List:
```
Comment (Thenerdstation) [inline tensornetwork/matrixproductstates/mps.py]: Give this a full name too.

File: tensornetwork/matrixproductstates/mps.py [hunk 8]
```diff
@@ -0,0 +1,564 @@
...
+          self.nodes[n - 1].tensor /= Z
+
+      self.center_position = site
+    return Z
+
+  def check_orthonormality(self, which: Text, site: int) -> Tensor:
+    """
+    Check orthonormality of tensor at site `site`.
+    Args:
+      which: if in ('l','left'): check left orthogonality
```
Comment (Thenerdstation) [inline tensornetwork/matrixproductstates/mps.py]: if which is 'l' or 'left'

File: tensornetwork/matrixproductstates/mps.py [hunk 9]
```diff
@@ -0,0 +1,564 @@
...
+  def __len__(self):
+    return len(self.nodes)
+
+  def apply_transfer_operator(self, site: int, direction: Union[Text, int],
+                              matrix: Tensor) -> Tensor:
+    """
+    Compute the action of the MPS transfer-operator at site `site`.
+    Args:
+      site (int): a site of the MPS
+      direction (str or int): if in (1, 'l', 'left'): compute the left-action
```
Comment (Thenerdstation) [inline tensornetwork/matrixproductstates/mps.py]: If direction is 1, 'l' or 'left' then...

File: tensornetwork/matrixproductstates/mps_class_test.py [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+      get_random_np((D, d, D), dtype) for _ in range(N - 2)
+  ] + [get_random_np((D, d, 1), dtype)]
+  mps = tensornetwork.mps.FiniteMPS(tensors, center_position=0, backend=backend)
+  gate = get_random_np((2, 2), dtype)
+  mps.apply_one_site_gate(gate, 5)
+  actual = np.transpose(np.tensordot(tensors[5], gate, ([1], [1])), (0, 2, 1))
+  np.testing.assert_allclose(mps.nodes[5].tensor, actual)
+
+
+@pytest.mark.parametrize("backend,dtype", [('numpy', np.float64),
```
Comment (Thenerdstation) [inline tensornetwork/matrixproductstates/mps_class_test.py]: I think you can avoid copying and pasting these by using pytest's fixures: https://pybit.es/pytest-fixtures.html

We use this already here https://github.com/google/TensorNetwork/blob/50849ce792a1dcdf2a445af83125d06a5c25df09/conftest.py#L21, You could do the same by adding a local fixure in this file.

File: tensornetwork/matrixproductstates/mps.py [hunk 11]
```diff
@@ -0,0 +1,564 @@
...
+    """
+    Shift FiniteMPS.center_position to `site`.
+    Args:
+      site: The site to which FiniteMPS.center_position should be shifted
+      normalize: If `True`, normalize matrices when shifting.
+    Returns:
+      Tensor: The norm of the tensor at FiniteMPS.center_position
+    """
+
+    if site >= len(self.nodes) or site < 0:
```
Comment (Thenerdstation) [inline tensornetwork/matrixproductstates/mps.py]: Could you add comments to this code? (And throughout the rest of this file.)

File: tensornetwork/matrixproductstates/mps_class_test.py (hunk 10 above)
Comment (mganahl) [reply tensornetwork/matrixproductstates/mps_class_test.py]: Are you suggesting to add a fixture to conftest?

Comment (mganahl) [reply tensornetwork/matrixproductstates/mps_class_test.py]: btw, is that typo there on purpose?

Comment (mganahl) [reply tensornetwork/matrixproductstates/mps_class_test.py]: I'm not very familiar with this. where is `backend_fixure` called?

Comment (Thenerdstation) [reply tensornetwork/matrixproductstates/mps_class_test.py]: No, you would add the fixture to this file and not conftest.py

And pytest just handles it magically. It's great!

Comment (mganahl) [reply tensornetwork/matrixproductstates/mps_class_test.py]: Allright, and how exactly do I define a fixture with the particular combination of backend-dtype arguments (not all backends are called for all dtypes)?

Comment (mganahl) [reply tensornetwork/matrixproductstates/mps_class_test.py]: @Thenerdstation added the fixture

File: mps/mps_test.py (hunk 3 above)
Comment (mganahl) [reply mps/mps_test.py]: I've added the seed to the initializer

File: mps/mps.py (hunk 4 above)
Comment (mganahl) [reply mps/mps.py]: renamed

File: tensornetwork/matrixproductstates/mps.py (hunk 7 above)
Comment (mganahl) [reply tensornetwork/matrixproductstates/mps.py]: yeah, a single letter is bit obscure

File: tensornetwork/matrixproductstates/mps.py (hunk 11 above)
Comment (mganahl) [reply tensornetwork/matrixproductstates/mps.py]: Done!

**Comment by mganahl** [comment]
Hey @Thenerdstation, this is good from my side

File: tensornetwork/matrixproductstates/mps.py [hunk 12]
```diff
@@ -0,0 +1,644 @@
...
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from tensornetwork.network import TensorNetwork
+import numpy as np
+from typing import Any, List, Optional, Text, Type, Union, Dict
+Tensor = Any
+
+
+class FiniteMPS(TensorNetwork):
```
Comment (Thenerdstation) [inline tensornetwork/matrixproductstates/mps.py]: I don't think it's a good idea to have FiniteMPS inherit TensorNetwork. I really doubt FiniteMPS should be able to do everything that TensorNetwork can do. I think it would be better to have the network as a property of the FiniteMPS.

Comment (mganahl) [reply tensornetwork/matrixproductstates/mps.py]: yes, I think I agree. Wasn't sure about this either, I wanted to try it to see how convenient it is.

Comment (mganahl) [reply tensornetwork/matrixproductstates/mps.py]: @Thenerdstation I removed inheritance from `TensorNetwork`

File: tensornetwork/matrixproductstates/mps.py [hunk 13]
```diff
@@ -0,0 +1,636 @@
...
+import tensornetwork as tn
+from typing import Any, List, Optional, Text, Type, Union, Dict, Sequence
+Tensor = Any
+
+
+class FiniteMPS:
+  """
+  An MPS class for finite systems.
+  `FiniteMPS` keeps track of the nodes of the network by storing them in a list
+  `FiniteMPS._nodes`. Any external changes to this list will potentially corrupt
```
Comment (Thenerdstation) [inline tensornetwork/matrixproductstates/mps.py]: Don't mention private `_node` properties in the class description.

File: tensornetwork/matrixproductstates/mps.py [hunk 14]
```diff
@@ -0,0 +1,636 @@
...
+      self.center_position = site
+    #return the norm of the last R tensor (useful for checks)
+    return Z
+
+  def check_orthonormality(self, which: Text, site: int) -> Tensor:
+    """
+    Check orthonormality of tensor at site `site`.
+    Args:
+      which: if 'l' or 'left': check left orthogonality
+             if 'r' or'right': check right orthogonality
```
Comment (Thenerdstation) [inline tensornetwork/matrixproductstates/mps.py]: Add space after "or"

File: tensornetwork/matrixproductstates/mps.py [hunk 15]
```diff
@@ -0,0 +1,636 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import numpy as np
+import tensornetwork as tn
```
Comment (Thenerdstation) [inline tensornetwork/matrixproductstates/mps.py]: Don't import `tensornetwork` directly. It can lead to circular import of we decided to make `mps` easier to access by putting it in `tensornetwork/__init__.py` file. Prefer to import everything you need directly.

**Comment by mganahl** [comment]
removed the import

File: tensornetwork/matrixproductstates/mps.py (hunk 13 above)
Comment (Thenerdstation) [reply tensornetwork/matrixproductstates/mps.py]: plz fix

File: tensornetwork/matrixproductstates/mps.py (hunk 14 above)
Comment (Thenerdstation) [reply tensornetwork/matrixproductstates/mps.py]: plz fix

**Comment by mganahl** [comment]
sorry, had forgotten about your earlier comments. Fixed this
